Sources/OpenAPIRuntime/Interface/HTTPBody.swift

```diff
@@ -0,0 +1,754 @@
+    ///   - length: The total lenght of the body.
+    ///   - iterationBehavior: The iteration behavior of the sequence, which
+    ///     indicates whether it can be iterated multiple times.
+    @inlinable public convenience init<Bytes: AsyncSequence>(
+        _ sequence: Bytes,
+        length: HTTPBody.Length,
+        iterationBehavior: IterationBehavior
+    ) where Bytes: Sendable, Bytes.Element: Sequence, Bytes.Element.Element == UInt8 {
+        self.init(
+            sequence.map { ArraySlice($0) },
+            length: length,
+            iterationBehavior: iterationBehavior
+        )
+    }
+}
+
+// MARK: - Consuming the body
+
+extension HTTPBody: AsyncSequence {
+    public typealias Element = ByteChunk
+    public typealias AsyncIterator = Iterator
+    public func makeAsyncIterator() -> AsyncIterator {
+        if iterationBehavior == .single {
+            lock.lock()
+            defer {
+                lock.unlock()
+            }
+            guard !locked_iteratorCreated else {
+                fatalError(
+                    "OpenAPIRuntime.HTTPBody attempted to create a second iterator, but the underlying sequence is only safe to be iterated once."
+                )
+            }
+            locked_iteratorCreated = true
+        }
+        return sequence.makeAsyncIterator()
+    }
+}
+
+extension HTTPBody {
+
+    /// An error thrown by the collecting initializer when the body contains more
+    /// than the maximum allowed number of bytes.
+    private struct TooManyBytesError: Error, CustomStringConvertible, LocalizedError {
+
+        /// The maximum number of bytes acceptable by the user.
+        let maxBytes: Int
+
+        var description: String {
+            "OpenAPIRuntime.HTTPBody contains more than the maximum allowed \(maxBytes) bytes."
+        }
+
+        var errorDescription: String? {
+            description
+        }
+    }
+
+    /// An error thrown by the collecting initializer when another iteration of
+    /// the body is not allowed.
+    private struct TooManyIterationsError: Error, CustomStringConvertible, LocalizedError {
+
+        var description: String {
+            "OpenAPIRuntime.HTTPBody attempted to create a second iterator, but the underlying sequence is only safe to be iterated once."
+        }
+
+        var errorDescription: String? {
+            description
+        }
+    }
+
+    /// Accumulates the full body in-memory into a single buffer
+    /// up to the provided maximum number of bytes and returns it.
+    /// - Parameters:
+    ///   - maxBytes: The maximum number of bytes this method is allowed
+    ///     to accumulate in memory before it throws an error.
+    /// - Throws: `TooManyBytesError` if the body contains more
+    ///   than `maxBytes`.
+    /// - Returns: A byte chunk containing all the accumulated bytes.
+    fileprivate func collect(upTo maxBytes: Int) async throws -> ByteChunk {
+
```

Comment:
Can you elaborate? I agree we should somehow take into account the known length of the body (when available), but how exactly? Should we `min(maxBytes, knownByteCount)` or something?

Reply:
The documentation says it throws `TooManyBytesError` if the body contains more than `maxBytes`.

If we have a known length we can fail early, rather than lazily, and we can then not taint the sequence (by iterating it, in the case it is `.single`).

```swift
if case let .known(knownBytes) = length {
    guard knownBytes < maxBytes else {
        throw TooManyBytesError(maxBytes: maxBytes)
    }
}
```

Reply:
Good idea, added https://github.com/apple/swift-openapi-runtime/pull/47/commits/72d241d0f22a9ac11482213b3febc0f8d24837b9

Reply:
Looks great, can we add a test for it?

Reply:
Added: https://github.com/apple/swift-openapi-runtime/pull/47/commits/07046738cf0929dd96339a0706e8f944507f2b62

Reply:
Awesome, thanks!